### PR TITLE
Add iap refund support

### PR DESCRIPTION
**Pull request by sesposito**
Add IAP hooks for incoming Apple and Google notifications to handle refunds.

### CODE REVIEW

File: server/core_subscription.go [hunk 1]
```diff
@@ -619,26 +736,107 @@ func appleNotificationHandler(logger *zap.Logger, db *sql.DB) http.HandlerFunc {
...
+						logger.Error("Error invoking Apple subscription refund runtime function", zap.Error(err))
+						w.WriteHeader(http.StatusOK)
+						return
+					}
+				}
+			}
+
+		} else {
+			// Notification regarding a purchase.
+			if strings.ToUpper(notification.NotificationType) == "REFUND" {
```
Comment (zyro) [inline server/core_subscription.go:799]: Maybe use some constants, instead of `"REFUND"` as a string here?

File: server/runtime.go [hunk 2]
```diff
@@ -261,6 +270,14 @@ func (e RuntimeExecutionMode) String() string {
...
 	case RuntimeExecutionModeLeaderboardReset:
 		return "leaderboard_reset"
+	case RuntimeExecutionModePurchaseNotificationApple:
+		return "purchase_notification_apple"
+	case RuntimeExecutionModeSubscriptionNotificationApple:
+		return "subscription_notification_apple"
+	case RuntimeExecutionModePurchaseNotificationGoogle:
+		return "purchase_notification_google"
+	case RuntimeExecutionModeSubscriptionNotificationGoogle:
+		return "subscription_notification_apple"
```
Comment (zyro) [inline server/runtime.go:280]: Typo, should be `_google`.

File: server/runtime_lua.go [hunk 3]
```diff
@@ -1986,6 +2142,7 @@ func (r *RuntimeLua) GetCallback(e RuntimeExecutionMode, key string) *lua.LFunct
 		return r.callbacks.TournamentReset
 	case RuntimeExecutionModeLeaderboardReset:
 		return r.callbacks.LeaderboardReset
+
```
Comment (zyro) [inline server/runtime_lua.go:2145]: Do we need the new execution modes added here?

**Review by zyro — CHANGES_REQUESTED** [review]

File: iap/iap.go [hunk 4]
```diff
@@ -471,6 +478,115 @@ func validateReceiptGoogleWithIDs(ctx context.Context, httpc *http.Client, token
 	}
 }
 
+/*type VoidedReceiptsGoogle struct {
```
[the remark this answers is not in the document]
Comment (sesposito) [reply iap/iap.go:481]: Some code that once used but not anymore, will remove

File: iap/iap.go [hunk 5]
```diff
@@ -462,6 +463,12 @@ func validateReceiptGoogleWithIDs(ctx context.Context, httpc *http.Client, token
 		}
 
 		return out, gr, buf, nil
+	case 403:
+		// TODO: refresh token and retry request.
+		if !retried {
+			// goto
```
[the remark this answers is not in the document]
Comment (sesposito) [reply iap/iap.go:469]: I was considering adding a retry mechanism but ended up scrapping it, will remove

File: server/core_purchase.go [hunk 6]
```diff
@@ -510,26 +562,35 @@ type storagePurchase struct {
...
 	statements := make([]string, 0, len(purchases))
-	params := make([]interface{}, 0, len(purchases)*7)
+	params := make([]interface{}, 0, len(purchases)*8)
 	transactionIDsToPurchase := make(map[string]*storagePurchase)
 	offset := 0
 	for _, purchase := range purchases {
-		statement := fmt.Sprintf("($%d, $%d, $%d, $%d, $%d, $%d, $%d)", offset+1, offset+2, offset+3, offset+4, offset+5, offset+6, offset+7)
-		offset += 7
+		t := time.Time{}
+		if purchase.refundTime == t {
```
[the remark this answers is not in the document]
Comment (sesposito) [reply server/core_purchase.go:583]: Nice, didn't know about this

**Review by zyro — APPROVED** [review]

**Pull request merged**